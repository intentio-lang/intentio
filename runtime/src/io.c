// For getline
#ifdef __STDC_ALLOC_LIB__
#define __STDC_WANT_LIB_EXT2__ 1
#else
#define _POSIX_C_SOURCE 200809L
#endif

#include "io.h"

#include <stdio.h>

#include "none.h"
#include "str.h"
#include "util.h"

IeoResult
ieo_println(IEO_NOTNULL IeoTerm *s)
{
  IeoResult e = IEO_SUCCT(s);

  if (IEO_ERR(ieo_is_string(s))) {
    IEO_TRY(e, ieo_str(s));
    return ieo_println(e.term);
  }

  const char *cstr = ieo_string_c_str(s);
  IEO_ASSERT(cstr != NULL);

  printf("%s\n", cstr);

  return e;
}

IeoResult
ieo_scanln()
{
  IEO_STATIC_STRING(FERR, "Error reading from standard input.");

  char *data = NULL;
  size_t len = 0;
  ssize_t nread = getline(&data, &len, stdin);

  if (nread < 0) {
    free(data);
    return IEO_FAILT(&FERR);
  }

  // Trim trailing \n
  if (nread > 0) {
    nread--;
  }

  IeoResult e = ieo_string_new(data, (size_t)nread);
  free(data);
  return e;
}
