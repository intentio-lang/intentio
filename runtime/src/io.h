#pragma once

#include "term.h"

IeoResult
ieo_println(IEO_NOTNULL IeoTerm *s);

IeoResult
ieo_scanln();
