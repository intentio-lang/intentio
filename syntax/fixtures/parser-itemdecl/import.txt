import math:sin as sinus
import math:cos
import tg as tangens
import ctg
import io:*
