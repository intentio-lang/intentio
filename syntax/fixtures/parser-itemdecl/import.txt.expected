[
  {
    "_itemDeclAnn": [],
    "_itemDeclKind": {
      "contents": {
        "_importDeclAnn": [],
        "_importDeclKind": {
          "contents": [
            {
              "_qidAnn": [],
              "_qidMod": "math",
              "_qidScope": "sin",
              "_qidSourcePos": [
                "test.ieo",
                0,
                0
              ]
            },
            {
              "_scopeIdAnn": [],
              "_scopeIdSourcePos": [
                "test.ieo",
                0,
                0
              ],
              "_unScopeId": "sinus"
            }
          ],
          "tag": "ImportQidAs"
        },
        "_importDeclSourcePos": [
          "test.ieo",
          0,
          0
        ]
      },
      "tag": "ImportItemDecl"
    },
    "_itemDeclSourcePos": [
      "test.ieo",
      0,
      0
    ]
  },
  {
    "_itemDeclAnn": [],
    "_itemDeclKind": {
      "contents": {
        "_importDeclAnn": [],
        "_importDeclKind": {
          "contents": {
            "_qidAnn": [],
            "_qidMod": "math",
            "_qidScope": "cos",
            "_qidSourcePos": [
              "test.ieo",
              0,
              0
            ]
          },
          "tag": "ImportQid"
        },
        "_importDeclSourcePos": [
          "test.ieo",
          0,
          0
        ]
      },
      "tag": "ImportItemDecl"
    },
    "_itemDeclSourcePos": [
      "test.ieo",
      0,
      0
    ]
  },
  {
    "_itemDeclAnn": [],
    "_itemDeclKind": {
      "contents": {
        "_importDeclAnn": [],
        "_importDeclKind": {
          "contents": [
            {
              "_modIdAnn": [],
              "_modIdSourcePos": [
                "test.ieo",
                0,
                0
              ],
              "_unModId": "tg"
            },
            {
              "_modIdAnn": [],
              "_modIdSourcePos": [
                "test.ieo",
                0,
                0
              ],
              "_unModId": "tangens"
            }
          ],
          "tag": "ImportIdAs"
        },
        "_importDeclSourcePos": [
          "test.ieo",
          0,
          0
        ]
      },
      "tag": "ImportItemDecl"
    },
    "_itemDeclSourcePos": [
      "test.ieo",
      0,
      0
    ]
  },
  {
    "_itemDeclAnn": [],
    "_itemDeclKind": {
      "contents": {
        "_importDeclAnn": [],
        "_importDeclKind": {
          "contents": {
            "_modIdAnn": [],
            "_modIdSourcePos": [
              "test.ieo",
              0,
              0
            ],
            "_unModId": "ctg"
          },
          "tag": "ImportId"
        },
        "_importDeclSourcePos": [
          "test.ieo",
          0,
          0
        ]
      },
      "tag": "ImportItemDecl"
    },
    "_itemDeclSourcePos": [
      "test.ieo",
      0,
      0
    ]
  },
  {
    "_itemDeclAnn": [],
    "_itemDeclKind": {
      "contents": {
        "_importDeclAnn": [],
        "_importDeclKind": {
          "contents": {
            "_modIdAnn": [],
            "_modIdSourcePos": [
              "test.ieo",
              0,
              0
            ],
            "_unModId": "io"
          },
          "tag": "ImportAll"
        },
        "_importDeclSourcePos": [
          "test.ieo",
          0,
          0
        ]
      },
      "tag": "ImportItemDecl"
    },
    "_itemDeclSourcePos": [
      "test.ieo",
      0,
      0
    ]
  }
]
